File: packages/sdk/src/resources/disputes.ts
/**
 * Disputes resource client.
 *
 * Maps to `/v1/disputes`. A dispute is opened against a confirmed payment,
 * evidence is submitted while it is `open`/`under_review`, and it is resolved
 * with an outcome (`won` / `lost` / `refunded`).
 */
import type { HttpClient, RequestOptions } from "../http-client.js";

/** Why a dispute was opened. */
export type DisputeReason = "fraud" | "not_received" | "duplicate" | "quality" | "unrecognized";

/** Lifecycle of a dispute. */
export type DisputeStatus = "open" | "under_review" | "won" | "lost" | "refunded";

/** The kind of evidence attached to a dispute. */
export type DisputeEvidenceKind =
  | "text"
  | "receipt"
  | "shipping"
  | "communication"
  | "url"
  | "file";

/** A single piece of evidence submitted for a dispute. */
export interface DisputeEvidence {
  id: string;
  kind: DisputeEvidenceKind;
  description: string;
  value: string;
  submittedAt: string;
}

/** A dispute against a confirmed payment, as returned by the API. */
export interface Dispute {
  id: string;
  paymentId: string;
  customerId: string;
  reason: DisputeReason;
  status: DisputeStatus;
  evidence: DisputeEvidence[];
  openedAt: string;
  updatedAt: string;
  resolvedAt?: string;
}

/** Input for {@link DisputesResource.open}. */
export interface OpenDisputeInput {
  paymentId: string;
  customerId: string;
  reason: DisputeReason;
}

/** Input for {@link DisputesResource.submitEvidence}. */
export interface SubmitEvidenceInput {
  kind: DisputeEvidenceKind;
  description: string;
  value: string;
}

/** Resolution outcome for a dispute. */
export type DisputeOutcome = "won" | "lost" | "refunded";

/** Client for dispute endpoints. */
export class DisputesResource {
  constructor(private readonly http: HttpClient) {}

  /** List disputes, optionally filtered by status. */
  list(status?: DisputeStatus, options?: RequestOptions): Promise<Dispute[]> {
    const qs = status ? `?status=${encodeURIComponent(status)}` : "";
    return this.http.get<Dispute[]>(`/v1/disputes${qs}`, options);
  }

  /** Open a dispute against a payment. */
  open(input: OpenDisputeInput, options?: RequestOptions): Promise<Dispute> {
    return this.http.post<Dispute>("/v1/disputes", input, options);
  }

  /** Retrieve a dispute by id. */
  retrieve(id: string, options?: RequestOptions): Promise<Dispute> {
    return this.http.get<Dispute>(`/v1/disputes/${encodeURIComponent(id)}`, options);
  }

  /** Submit a piece of evidence for a dispute. */
  submitEvidence(id: string, input: SubmitEvidenceInput, options?: RequestOptions): Promise<Dispute> {
    return this.http.post<Dispute>(
      `/v1/disputes/${encodeURIComponent(id)}/evidence`,
      input,
      options,
    );
  }

  /** Resolve a dispute with an outcome. */
  resolve(id: string, outcome: DisputeOutcome, options?: RequestOptions): Promise<Dispute> {
    return this.http.post<Dispute>(
      `/v1/disputes/${encodeURIComponent(id)}/resolve`,
      { outcome },
      options,
    );
  }
}

File: sdks/go/disputes.go
package settlekit

import (
	"context"
	"net/http"
)

// DisputeEvidence is a single piece of evidence attached to a dispute.
type DisputeEvidence struct {
	ID          string `json:"id"`
	Kind        string `json:"kind"`
	Description string `json:"description"`
	Value       string `json:"value"`
	SubmittedAt string `json:"submittedAt"`
}

// Dispute is a dispute opened against a confirmed payment.
type Dispute struct {
	ID         string            `json:"id"`
	PaymentID  string            `json:"paymentId"`
	CustomerID string            `json:"customerId"`
	Reason     string            `json:"reason"`
	Status     string            `json:"status"`
	Evidence   []DisputeEvidence `json:"evidence"`
	OpenedAt   string            `json:"openedAt"`
	UpdatedAt  string            `json:"updatedAt"`
	ResolvedAt string            `json:"resolvedAt,omitempty"`
}

// OpenDisputeInput is the body for OpenDispute (POST /v1/disputes). Reason is
// one of: "fraud", "not_received", "duplicate", "quality", "unrecognized".
type OpenDisputeInput struct {
	PaymentID  string `json:"paymentId"`
	CustomerID string `json:"customerId"`
	Reason     string `json:"reason"`
}

// OpenDispute opens a dispute against a payment. POST /v1/disputes.
func (c *Client) OpenDispute(ctx context.Context, in OpenDisputeInput) (*Dispute, error) {
	var out Dispute
	if err := c.do(ctx, http.MethodPost, "/v1/disputes", in, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// ListDisputes lists disputes, optionally filtered by status (pass "" for all).
// GET /v1/disputes[?status=].
func (c *Client) ListDisputes(ctx context.Context, status string) ([]Dispute, error) {
	q := newQuery()
	q.add("status", status)
	var out []Dispute
	if err := c.do(ctx, http.MethodGet, "/v1/disputes"+q.encode(), nil, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// GetDispute fetches a dispute by id. GET /v1/disputes/:id.
func (c *Client) GetDispute(ctx context.Context, id string) (*Dispute, error) {
	var out Dispute
	if err := c.do(ctx, http.MethodGet, "/v1/disputes/"+id, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// SubmitEvidenceInput is the body for SubmitDisputeEvidence. Kind is one of:
// "text", "receipt", "shipping", "communication", "url", "file".
type SubmitEvidenceInput struct {
	Kind        string `json:"kind"`
	Description string `json:"description"`
	Value       string `json:"value"`
}

// SubmitDisputeEvidence attaches evidence to a dispute.
// POST /v1/disputes/:id/evidence.
func (c *Client) SubmitDisputeEvidence(ctx context.Context, id string, in SubmitEvidenceInput) (*Dispute, error) {
	var out Dispute
	if err := c.do(ctx, http.MethodPost, "/v1/disputes/"+id+"/evidence", in, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// ResolveDispute resolves a dispute with an outcome ("won", "lost", "refunded").
// POST /v1/disputes/:id/resolve.
func (c *Client) ResolveDispute(ctx context.Context, id, outcome string) (*Dispute, error) {
	var out Dispute
	body := map[string]string{"outcome": outcome}
	if err := c.do(ctx, http.MethodPost, "/v1/disputes/"+id+"/resolve", body, &out); err != nil {
		return nil, err
	}
	return &out, nil
}
